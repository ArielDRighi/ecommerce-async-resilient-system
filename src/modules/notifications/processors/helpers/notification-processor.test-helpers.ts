import { Job } from 'bull';
import { NotificationsService } from '../../notifications.service';
import { NotificationPriority } from '../../enums';

/**
 * Create mock NotificationsService
 */
export function mockNotificationsService(): jest.Mocked<NotificationsService> {
  return {
    sendOrderConfirmation: jest.fn().mockResolvedValue(undefined),
    sendPaymentFailure: jest.fn().mockResolvedValue(undefined),
    sendShippingUpdate: jest.fn().mockResolvedValue(undefined),
    sendWelcomeEmail: jest.fn().mockResolvedValue(undefined),
  } as unknown as jest.Mocked<NotificationsService>;
}

/**
 * Create mock job for testing
 */
export function createMockJob(
  id: string,
  type: 'order-confirmation' | 'payment-failure' | 'shipping-update' | 'welcome',
  data: any, // eslint-disable-line @typescript-eslint/no-explicit-any
  priority: NotificationPriority = NotificationPriority.NORMAL,
  userId: string = 'user-123',
): Job {
  return {
    id,
    data: {
      type,
      data,
      priority,
      userId,
    },
  } as Job;
}

/**
 * Assert successful job result
 */
export function expectSuccessfulJobResult(result: any, jobId: string) { // eslint-disable-line @typescript-eslint/no-explicit-any
  expect(result).toEqual({
    success: true,
    jobId,
  });
}

/**
 * Assert job result structure
 */
export function expectValidJobResultStructure(result: any) { // eslint-disable-line @typescript-eslint/no-explicit-any
  expect(result).toHaveProperty('success');
  expect(result).toHaveProperty('jobId');
  expect(typeof result.success).toBe('boolean');
  expect(typeof result.jobId).toBe('string');
}
