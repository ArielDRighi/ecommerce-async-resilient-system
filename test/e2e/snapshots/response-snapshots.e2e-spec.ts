import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import { AppModule } from '../../../src/app.module';
import request from 'supertest';
import { generateTestEmail, generateTestSKU } from '../../helpers/mock-data';

/**
 * Response Snapshots E2E Tests
 * Category: Snapshot Tests
 * Purpose: Validate response structure consistency using Jest snapshots
 */
describe('Response Snapshots (E2E)', () => {
  let app: INestApplication;
  let userToken: string;
  let productId: string;
  let orderId: string;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(
      new ValidationPipe({
        whitelist: true,
        forbidNonWhitelisted: true,
        transform: true,
      }),
    );
    await app.init();

    // Setup test data
    const userResponse = await request(app.getHttpServer()).post('/auth/register').send({
      email: generateTestEmail(),
      password: 'TestPassword123!',
      firstName: 'Snapshot',
      lastName: 'User',
    });
    userToken = userResponse.body.data.accessToken;

    const productResponse = await request(app.getHttpServer())
      .post('/products')
      .set('Authorization', `Bearer ${userToken}`)
      .send({
        name: 'Snapshot Test Product',
        description: 'Product for snapshot testing',
        price: 99.99,
        sku: generateTestSKU(),
        brand: 'TestBrand',
        weight: 1.5,
      });
    productId = productResponse.body.data.id;

    const orderResponse = await request(app.getHttpServer())
      .post('/orders')
      .set('Authorization', `Bearer ${userToken}`)
      .send({
        items: [
          {
            productId: productId,
            quantity: 2,
          },
        ],
      });
    orderId = orderResponse.body.data.id;
  });

  afterAll(async () => {
    if (app) {
      await app.close();
    }
  });

  describe('Authentication Response Snapshots', () => {
    it('should match registration response structure', async () => {
      const response = await request(app.getHttpServer())
        .post('/auth/register')
        .send({
          email: generateTestEmail(),
          password: 'NewUser123!',
          firstName: 'New',
          lastName: 'User',
        })
        .expect(201);

      // Normalize dynamic fields
      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        data: {
          ...response.body.data,
          accessToken: expect.any(String),
          refreshToken: expect.any(String),
          user: {
            ...response.body.data.user,
            id: expect.any(String),
            email: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          },
        },
      };

      expect(normalized).toMatchSnapshot();
    });

    it('should match login response structure', async () => {
      const email = generateTestEmail();
      const password = 'LoginUser123!';

      await request(app.getHttpServer()).post('/auth/register').send({
        email,
        password,
        firstName: 'Login',
        lastName: 'User',
      });

      const response = await request(app.getHttpServer())
        .post('/auth/login')
        .send({
          email,
          password,
        })
        .expect(200);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        data: {
          ...response.body.data,
          accessToken: expect.any(String),
          refreshToken: expect.any(String),
          user: {
            ...response.body.data.user,
            id: expect.any(String),
            email: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            lastLoginAt: expect.any(String),
          },
        },
      };

      expect(normalized).toMatchSnapshot();
    });
  });

  describe('Product Response Snapshots', () => {
    it('should match product detail response structure', async () => {
      const response = await request(app.getHttpServer()).get(`/products/${productId}`).expect(200);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        path: expect.any(String),
        data: {
          ...response.body.data,
          id: expect.any(String),
          sku: expect.any(String),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
        },
      };

      expect(normalized).toMatchSnapshot();
    });

    it('should match product list response structure', async () => {
      const response = await request(app.getHttpServer())
        .get('/products')
        .query({ page: 1, limit: 10 })
        .expect(200);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        data: {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          data: response.body.data.data.map((product: any) => ({
            ...product,
            id: expect.any(String),
            sku: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          })),
          meta: {
            ...response.body.data.meta,
            totalItems: expect.any(Number),
          },
        },
      };

      expect(normalized).toMatchSnapshot();
    });

    it('should match product creation response structure', async () => {
      const response = await request(app.getHttpServer())
        .post('/products')
        .set('Authorization', `Bearer ${userToken}`)
        .send({
          name: 'New Snapshot Product',
          description: 'Product created for snapshot',
          price: 149.99,
          sku: generateTestSKU(),
        })
        .expect(201);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        data: {
          ...response.body.data,
          id: expect.any(String),
          sku: expect.any(String),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
        },
      };

      expect(normalized).toMatchSnapshot();
    });
  });

  describe('Order Response Snapshots', () => {
    it('should match order detail response structure', async () => {
      const response = await request(app.getHttpServer())
        .get(`/orders/${orderId}`)
        .set('Authorization', `Bearer ${userToken}`)
        .expect(200);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        path: expect.any(String),
        data: {
          ...response.body.data,
          id: expect.any(String),
          userId: expect.any(String),
          idempotencyKey: expect.any(String),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          items: response.body.data.items?.map((item: any) => ({
            ...item,
            id: expect.any(String),
            productId: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          })),
        },
      };

      expect(normalized).toMatchSnapshot();
    });

    it('should match order list response structure', async () => {
      const response = await request(app.getHttpServer())
        .get('/orders')
        .set('Authorization', `Bearer ${userToken}`)
        .query({ page: 1, limit: 10 })
        .expect(200);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        data: {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          data:
            response.body.data.data?.map((order: any) => ({
              ...order,
              id: expect.any(String),
              userId: expect.any(String),
              idempotencyKey: expect.any(String),
              createdAt: expect.any(String),
              updatedAt: expect.any(String),
            })) || [],
          meta: {
            ...response.body.data.meta,
            totalItems: expect.any(Number),
          },
        },
      };

      expect(normalized).toMatchSnapshot();
    });

    it('should match order creation response structure', async () => {
      const response = await request(app.getHttpServer())
        .post('/orders')
        .set('Authorization', `Bearer ${userToken}`)
        .send({
          items: [
            {
              productId: productId,
              quantity: 1,
            },
          ],
        })
        .expect(202);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        data: {
          ...response.body.data,
          id: expect.any(String),
          userId: expect.any(String),
          idempotencyKey: expect.any(String),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          items:
            response.body.data.items?.map((item: any) => ({
              ...item,
              id: expect.any(String),
              productId: expect.any(String),
            })) || [],
        },
      };

      expect(normalized).toMatchSnapshot();
    });
  });

  describe('User Response Snapshots', () => {
    it('should match user profile response structure', async () => {
      const response = await request(app.getHttpServer())
        .get('/users/profile')
        .set('Authorization', `Bearer ${userToken}`)
        .expect(200);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        data: {
          ...response.body.data,
          id: expect.any(String),
          email: expect.any(String),
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
          lastLoginAt: expect.any(String),
        },
      };

      expect(normalized).toMatchSnapshot();
    });
  });

  describe('Error Response Snapshots', () => {
    it('should match 400 Bad Request error structure', async () => {
      const response = await request(app.getHttpServer())
        .post('/products')
        .set('Authorization', `Bearer ${userToken}`)
        .send({
          name: 'Invalid Product',
          price: -10,
        })
        .expect(400);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        correlationId: expect.any(String),
        error: {
          ...response.body.error,
          message: expect.any(String),
        },
      };

      expect(normalized).toMatchSnapshot();
    });

    it('should match 401 Unauthorized error structure', async () => {
      const response = await request(app.getHttpServer()).get('/users/me').expect(401);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        error: {
          ...response.body.error,
          message: expect.any(String),
        },
      };

      expect(normalized).toMatchSnapshot();
    });

    it('should match 404 Not Found error structure', async () => {
      const response = await request(app.getHttpServer())
        .get('/products/00000000-0000-0000-0000-000000000000')
        .expect(404);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        correlationId: expect.any(String),
        error: {
          ...response.body.error,
          message: expect.any(String),
        },
      };

      expect(normalized).toMatchSnapshot();
    });
  });

  describe('Pagination Response Snapshots', () => {
    it('should match pagination metadata structure', async () => {
      const response = await request(app.getHttpServer())
        .get('/products')
        .query({ page: 2, limit: 5 })
        .expect(200);

      const normalized = {
        meta: {
          ...response.body.data.meta,
          totalItems: expect.any(Number),
        },
      };

      expect(normalized).toMatchSnapshot();
    });
  });

  describe('Health Check Response Snapshots', () => {
    it('should match health check response structure', async () => {
      const response = await request(app.getHttpServer()).get('/health').expect(200);

      const normalized = {
        ...response.body,
        timestamp: expect.any(String),
        uptime: expect.any(Number),
        memory: {
          used: expect.any(Number),
          total: expect.any(Number),
        },
      };

      expect(normalized).toMatchSnapshot();
    });
  });
});
