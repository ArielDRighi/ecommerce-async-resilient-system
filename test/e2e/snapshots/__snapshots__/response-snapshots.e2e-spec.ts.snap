// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response Snapshots (E2E) Authentication Response Snapshots should match login response structure 1`] = `
{
  "data": {
    "accessToken": Any<String>,
    "expiresIn": 86400,
    "refreshToken": Any<String>,
    "tokenType": "Bearer",
    "user": {
      "createdAt": Any<String>,
      "dateOfBirth": null,
      "email": Any<String>,
      "emailVerifiedAt": null,
      "firstName": "Login",
      "fullName": "Login User",
      "id": Any<String>,
      "isActive": true,
      "language": "en",
      "lastLoginAt": Any<String>,
      "lastName": "User",
      "phoneNumber": null,
      "timezone": "UTC",
      "updatedAt": Any<String>,
    },
  },
  "message": "Success",
  "path": "/auth/login",
  "statusCode": 200,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Authentication Response Snapshots should match registration response structure 1`] = `
{
  "data": {
    "accessToken": Any<String>,
    "expiresIn": 86400,
    "refreshToken": Any<String>,
    "tokenType": "Bearer",
    "user": {
      "createdAt": Any<String>,
      "dateOfBirth": null,
      "email": Any<String>,
      "emailVerifiedAt": null,
      "firstName": "New",
      "fullName": "New User",
      "id": Any<String>,
      "isActive": true,
      "language": "en",
      "lastLoginAt": null,
      "lastName": "User",
      "phoneNumber": null,
      "timezone": "UTC",
      "updatedAt": Any<String>,
    },
  },
  "message": "Created successfully",
  "path": "/auth/register",
  "statusCode": 201,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Error Response Snapshots should match 400 Bad Request error structure 1`] = `
{
  "correlationId": Any<String>,
  "error": {
    "0": "B",
    "1": "A",
    "10": "T",
    "2": "D",
    "3": "_",
    "4": "R",
    "5": "E",
    "6": "Q",
    "7": "U",
    "8": "E",
    "9": "S",
    "message": Any<String>,
  },
  "message": [
    "Price must be at least $0.01",
    "Price must be greater than 0",
    "SKU can only contain uppercase letters, numbers, hyphens, and underscores",
    "SKU must not exceed 100 characters",
    "SKU must be at least 3 characters long",
    "SKU is required",
    "SKU must be a string",
  ],
  "method": "POST",
  "path": "/products",
  "statusCode": 400,
  "success": false,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Error Response Snapshots should match 401 Unauthorized error structure 1`] = `
{
  "error": {
    "0": "U",
    "1": "N",
    "10": "E",
    "11": "D",
    "2": "A",
    "3": "U",
    "4": "T",
    "5": "H",
    "6": "O",
    "7": "R",
    "8": "I",
    "9": "Z",
    "message": Any<String>,
  },
  "message": "Unauthorized",
  "method": "GET",
  "path": "/users/me",
  "statusCode": 401,
  "success": false,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Error Response Snapshots should match 404 Not Found error structure 1`] = `
{
  "correlationId": Any<String>,
  "error": {
    "0": "N",
    "1": "O",
    "2": "T",
    "3": "_",
    "4": "F",
    "5": "O",
    "6": "U",
    "7": "N",
    "8": "D",
    "message": Any<String>,
  },
  "message": "Product with ID 00000000-0000-0000-0000-000000000000 not found",
  "method": "GET",
  "path": "/products/00000000-0000-0000-0000-000000000000",
  "statusCode": 404,
  "success": false,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Health Check Response Snapshots should match health check response structure 1`] = `
{
  "details": {
    "database": {
      "status": "up",
    },
    "memory_heap": {
      "status": "up",
    },
    "memory_rss": {
      "status": "up",
    },
    "storage": {
      "status": "up",
    },
  },
  "error": {},
  "info": {
    "database": {
      "status": "up",
    },
    "memory_heap": {
      "status": "up",
    },
    "memory_rss": {
      "status": "up",
    },
    "storage": {
      "status": "up",
    },
  },
  "memory": {
    "total": Any<Number>,
    "used": Any<Number>,
  },
  "status": "ok",
  "timestamp": Any<String>,
  "uptime": Any<Number>,
}
`;

exports[`Response Snapshots (E2E) Order Response Snapshots should match order creation response structure 1`] = `
{
  "data": {
    "createdAt": Any<String>,
    "currency": "USD",
    "id": Any<String>,
    "idempotencyKey": Any<String>,
    "items": [
      {
        "id": Any<String>,
        "productId": Any<String>,
        "productName": "Snapshot Test Product",
        "quantity": 1,
        "totalPrice": 99.99,
        "unitPrice": "99.99",
      },
    ],
    "status": "PENDING",
    "totalAmount": 99.99,
    "updatedAt": Any<String>,
    "userId": Any<String>,
  },
  "message": "Accepted",
  "path": "/orders",
  "statusCode": 202,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Order Response Snapshots should match order detail response structure 1`] = `
{
  "data": {
    "createdAt": Any<String>,
    "currency": "USD",
    "id": Any<String>,
    "idempotencyKey": Any<String>,
    "items": [
      {
        "createdAt": Any<String>,
        "id": Any<String>,
        "productId": Any<String>,
        "productName": "Snapshot Test Product",
        "quantity": 2,
        "totalPrice": "199.98",
        "unitPrice": "99.99",
        "updatedAt": Any<String>,
      },
    ],
    "status": "PENDING",
    "totalAmount": "199.98",
    "updatedAt": Any<String>,
    "userId": Any<String>,
  },
  "message": "Success",
  "path": Any<String>,
  "statusCode": 200,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Order Response Snapshots should match order list response structure 1`] = `
{
  "data": {
    "data": [],
    "meta": {
      "totalItems": Any<Number>,
    },
  },
  "message": "Success",
  "path": "/orders?page=1&limit=10",
  "statusCode": 200,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Pagination Response Snapshots should match pagination metadata structure 1`] = `
{
  "meta": {
    "hasNext": false,
    "hasPrev": true,
    "limit": 5,
    "page": 2,
    "total": 2,
    "totalItems": Any<Number>,
    "totalPages": 1,
  },
}
`;

exports[`Response Snapshots (E2E) Product Response Snapshots should match product creation response structure 1`] = `
{
  "data": {
    "attributes": null,
    "brand": null,
    "compareAtPrice": null,
    "costPrice": null,
    "createdAt": Any<String>,
    "deletedAt": null,
    "description": "Product created for snapshot",
    "discountPercentage": 0,
    "id": Any<String>,
    "images": null,
    "isActive": true,
    "isOnSale": false,
    "minimumStock": 0,
    "name": "New Snapshot Product",
    "price": "149.99",
    "profitMargin": 0,
    "sku": Any<String>,
    "tags": null,
    "trackInventory": true,
    "updatedAt": Any<String>,
    "weight": null,
  },
  "message": "Created successfully",
  "path": "/products",
  "statusCode": 201,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Product Response Snapshots should match product detail response structure 1`] = `
{
  "data": {
    "attributes": null,
    "brand": "TestBrand",
    "compareAtPrice": null,
    "costPrice": null,
    "createdAt": Any<String>,
    "deletedAt": null,
    "description": "Product for snapshot testing",
    "discountPercentage": 0,
    "id": Any<String>,
    "images": null,
    "isActive": true,
    "isOnSale": false,
    "minimumStock": 0,
    "name": "Snapshot Test Product",
    "price": "99.99",
    "profitMargin": 0,
    "sku": Any<String>,
    "tags": null,
    "trackInventory": true,
    "updatedAt": Any<String>,
    "weight": "1.500",
  },
  "message": "Success",
  "path": Any<String>,
  "statusCode": 200,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) Product Response Snapshots should match product list response structure 1`] = `
{
  "data": {
    "data": [
      {
        "attributes": null,
        "brand": "TestBrand",
        "compareAtPrice": null,
        "costPrice": null,
        "createdAt": Any<String>,
        "deletedAt": null,
        "description": "Product for snapshot testing",
        "discountPercentage": 0,
        "id": Any<String>,
        "images": null,
        "isActive": true,
        "isOnSale": false,
        "minimumStock": 0,
        "name": "Snapshot Test Product",
        "price": "99.99",
        "profitMargin": 0,
        "sku": Any<String>,
        "tags": null,
        "trackInventory": true,
        "updatedAt": Any<String>,
        "weight": "1.500",
      },
    ],
    "meta": {
      "hasNext": false,
      "hasPrev": false,
      "limit": 10,
      "page": 1,
      "total": 1,
      "totalItems": Any<Number>,
      "totalPages": 1,
    },
  },
  "message": "Success",
  "path": "/products?page=1&limit=10",
  "statusCode": 200,
  "success": true,
  "timestamp": Any<String>,
}
`;

exports[`Response Snapshots (E2E) User Response Snapshots should match user profile response structure 1`] = `
{
  "data": {
    "createdAt": Any<String>,
    "dateOfBirth": null,
    "email": Any<String>,
    "emailVerifiedAt": null,
    "firstName": "Snapshot",
    "fullName": "Snapshot User",
    "id": Any<String>,
    "isActive": true,
    "language": "en",
    "lastLoginAt": Any<String>,
    "lastName": "User",
    "phoneNumber": null,
    "timezone": "UTC",
    "updatedAt": Any<String>,
  },
  "message": "Success",
  "path": "/users/profile",
  "statusCode": 200,
  "success": true,
  "timestamp": Any<String>,
}
`;
